File: python/tunnel_test.py
#!/usr/bin/env python3

import errno
import pathlib
import socket
import sys
import threading
import typing

import saq.pqc.sandwich.proto.api.v1.configuration_pb2 as SandwichAPI
import saq.pqc.sandwich.proto.api.v1.encoding_format_pb2 as EncodingFormat
import saq.pqc.sandwich.proto.sandwich_pb2 as SandwichProto
import saq.pqc.sandwich.proto.tunnel_pb2 as SandwichTunnelProto
import saq.pqc.sandwich.python.errors as errors
import saq.pqc.sandwich.python.io as SandwichIO
from saq.pqc.sandwich.python.sandwich import Context, Sandwich, Tunnel

_LISTENING_ADDRESS = "127.0.0.1"
_LISTENING_PORT = 1339

_PING_MSG = b"PING"
_PONG_MSG = b"PONG"

_CERT_PATH = "testdata/cert.pem"
_KEY_PATH = "testdata/key.pem"

_DEFAULT_KEM = "kyber512"


class Socket(SandwichIO.IO):
    def __init__(self, sock: socket.socket):
        self._sock = sock

    _ERRNO_EXCEPTIONS_MAP = (
        (
            (
                errno.EINPROGRESS,
                errno.EINTR,
            ),
            errors.IOInProgressException,
        ),
        (
            (
                errno.EAGAIN,
                errno.EWOULDBLOCK,
            ),
            errors.IOWouldBlockException,
        ),
        (
            (
                errno.ENOTSOCK,
                errno.EPROTOTYPE,
                errno.EBADF,
            ),
            errors.IOInvalidException,
        ),
        (
            (
                errno.EACCES,
                errno.EPERM,
                errno.ETIMEDOUT,
                errno.ENETUNREACH,
                errno.ECONNREFUSED,
            ),
            errors.IORefusedException,
        ),
    )

    def _errno_to_exception(self, err):
        for val in Socket._ERRNO_EXCEPTIONS_MAP:
            if err in val[0]:
                return val[1]()
        return errors.IOUnknownException()

    def read(self, n, tunnel_state: SandwichTunnelProto.State) -> bytes:
        try:
            return self._sock.recv(n)
        except OSError as e:
            raise self._errno_to_exception(e.errno)

    def write(self, buf, tunnel_state: SandwichTunnelProto.State) -> int:
        try:
            return self._sock.send(buf)
        except Exception as e:
            raise self._errno_to_exception(e.errno)

    def close(self):
        self._sock.close()


def create_server_conf(s: Sandwich) -> Context:
    """Creates the configuration for the server.

    Returns:
        Configuration for the server.
    """
    conf = SandwichAPI.Configuration()
    conf.protocol = SandwichAPI.PROTO_TLS_13
    conf.impl = SandwichAPI.IMPL_OPENSSL1_1_1_OQS

    conf.server.tls.common_options.kem.append(_DEFAULT_KEM)
    conf.server.tls.certificate.static.data.filename = _CERT_PATH
    conf.server.tls.certificate.static.format = EncodingFormat.ENCODING_FORMAT_PEM

    conf.server.tls.private_key.static.data.filename = _KEY_PATH
    conf.server.tls.private_key.static.format = EncodingFormat.ENCODING_FORMAT_PEM

    return Context(s, conf)


def create_client_conf(s: Sandwich) -> Context:
    """Creates the configuration for the client.

    Returns:
        Configuration for the client.
    """
    conf = SandwichAPI.Configuration()
    conf.protocol = SandwichAPI.PROTO_TLS_13
    conf.impl = SandwichAPI.IMPL_OPENSSL1_1_1_OQS

    conf.client.tls.common_options.kem.append(_DEFAULT_KEM)

    cert = conf.client.tls.trusted_certificates.add().static
    cert.data.filename = _CERT_PATH
    cert.format = EncodingFormat.ENCODING_FORMAT_PEM

    return Context(s, conf)


def create_ios() -> (SandwichIO.IO, SandwichIO.IO):
    s1, s2 = socket.socketpair(
        family=socket.AF_UNIX, type=socket.SOCK_STREAM | socket.SOCK_NONBLOCK
    )
    return Socket(s1), Socket(s2)


def main():
    s = Sandwich()

    client_conf = create_client_conf(s)
    assert client_conf != None

    server_conf = create_server_conf(s)
    assert server_conf != None

    client_io, server_io = create_ios()
    assert client_io != None
    assert server_io != None

    server = Tunnel(server_conf, server_io)
    assert server != None

    client = Tunnel(client_conf, client_io)
    assert client != None

    e = None
    try:
        client.handshake()
        assert False, "expected  WANT_READ, got None"
    except errors.HandshakeException as e:
        assert isinstance(
            e, errors.HandshakeWantReadException
        ), f"expected WANT_READ, got {e}"
    except Exception as e:
        assert False, f"expected Tunnel.HandshakeWantReadException, got {e}"

    e = None
    try:
        server.handshake()
        assert False, "expected  WANT_READ, got None"
    except errors.HandshakeException as e:
        assert isinstance(
            e, errors.HandshakeWantReadException
        ), f"expected WANT_READ, got {e}"
    except Exception as e:
        assert False, f"expected Tunnel.HandshakeWantReadException, got {e}"

    e = None
    try:
        client.handshake()
    except Exception as e:
        assert False, f"expected no error, got {e}"

    e = None
    try:
        server.handshake()
    except Exception as e:
        assert False, f"expected no error, got {e}"

    state = client.state()
    assert (
        state == client.State.STATE_HANDSHAKE_DONE
    ), f"Expected state HANDSHAKE_DONE, got {state}"

    state = server.state()
    assert (
        state == client.State.STATE_HANDSHAKE_DONE
    ), f"Expected state HANDSHAKE_DONE, got {state}"

    w = 0
    try:
        w = client.write(_PING_MSG)
    except errors.RecordPlaneException as e:
        assert False, f"expected no error, got {e}"
    assert w == len(_PING_MSG), f"Expected {len(_PING_MSG)} bytes written, got {w}"

    try:
        data = server.read(len(_PING_MSG))
    except errors.RecordPlaneException as e:
        assert False, f"expected no error, got {e}"
    assert len(data) == len(_PING_MSG), f"Expected {len(_PING_MSG)} bytes read, got {w}"
    assert data == _PING_MSG, f"Expected msg {_PING_MSG} from client, got {data}"

    w = 0
    try:
        w = server.write(_PONG_MSG)
    except errors.RecordPlaneException as e:
        assert False, f"expected no error, got {e}"
    assert w == len(_PONG_MSG), f"Expected {len(_PING_MSG)} bytes written, got {w}"

    try:
        data = client.read(len(_PONG_MSG))
    except errors.RecordPlaneException as e:
        assert False, f"expected no error, got {e}"
    assert len(data) == len(_PONG_MSG), f"Expected {len(_PING_MSG)} bytes read, got {w}"
    assert data == _PONG_MSG, f"Expected msg {_PING_MSG} from client, got {data}"

    client.close()
    server.close()


if __name__ == "__main__":
    main()

File: rust/tunnel_test.rs
// Copyright 2022 SandboxAQ
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

extern crate api_rust_proto as SandwichAPI;
extern crate protobuf;
extern crate sandwich;
extern crate sandwich_rust_proto as SandwichProto;

/// A socket.
type IOBufferRC = std::rc::Rc<std::cell::RefCell<IOBuffer>>;
struct IOBuffer(Vec<u8>, Option<IOBufferRC>);

/// Implements sandwich::IO for IOBuffer.
impl sandwich::IO for IOBuffer {
    fn read(
        &mut self,
        b: &mut [u8],
        _: sandwich_rust_proto::State,
    ) -> sandwich::io::IOResult<usize> {
        if self.0.is_empty() {
            return Err(sandwich::errors::IOError::new(
                SandwichProto::IOError::IOERROR_WOULD_BLOCK,
            ));
        } else {
            let min = std::cmp::min(b.len(), self.0.len());
            b.clone_from_slice(&self.0[..min]);
            self.0.drain(..min);
            Ok(min)
        }
    }

    fn write(&mut self, b: &[u8], _: sandwich_rust_proto::State) -> sandwich::io::IOResult<usize> {
        self.1
            .as_ref()
            .unwrap()
            .as_ref()
            .borrow_mut()
            .0
            .extend_from_slice(b);
        Ok(b.len())
    }

    fn close(&mut self) {
        self.0.clear();
    }
}

#[cfg(test)]
mod tests {
    /// Path to the PEM-encoded certificate.
    static PEM_CERT_PATH: &str = "testdata/cert.pem";

    /// Path to the PEM-encoded private key.
    static PEM_KEY_PATH: &str = "testdata/key.pem";

    /// Default KEM.
    static DEFAULT_KEM: &str = "kyber512";

    /// Ping message.
    static PING_MSG: [u8; 4] = [0x50, 0x49, 0x4e, 0x47];

    /// Pong message.
    static PONG_MSG: [u8; 4] = [0x50, 0x4f, 0x4e, 0x47];

    /// Creates the Context for the client.
    fn create_client() -> Result<sandwich::Context, sandwich::errors::GlobalError> {
        let mut conf = SandwichAPI::Configuration::new();
        conf.protocol = SandwichAPI::Protocol::PROTO_TLS_13;
        conf.field_impl = SandwichAPI::Implementation::IMPL_OPENSSL1_1_1_OQS;
        let tls_client = conf.mut_client().mut_tls();
        tls_client
            .mut_common_options()
            .mut_kem()
            .push(DEFAULT_KEM.to_string());

        tls_client.mut_trusted_certificates().push({
            let mut cert = SandwichAPI::Certificate::new();
            let mut asn1 = cert.mut_field_static();
            asn1.mut_data().set_filename(PEM_CERT_PATH.to_string());
            asn1.format = SandwichAPI::encoding_format::ASN1EncodingFormat::ENCODING_FORMAT_PEM;
            cert
        });

        sandwich::Context::try_from(&conf)
    }

    /// Creates the Context for the server.
    fn create_server() -> Result<sandwich::Context, sandwich::errors::GlobalError> {
        let mut conf = SandwichAPI::Configuration::new();
        conf.protocol = SandwichAPI::Protocol::PROTO_TLS_13;
        conf.field_impl = SandwichAPI::Implementation::IMPL_OPENSSL1_1_1_OQS;
        let tls_server = conf.mut_server().mut_tls();
        tls_server
            .mut_common_options()
            .mut_kem()
            .push(DEFAULT_KEM.to_string());

        tls_server.set_certificate({
            let mut cert = SandwichAPI::Certificate::new();
            let mut asn1 = cert.mut_field_static();
            asn1.mut_data().set_filename(PEM_CERT_PATH.to_string());
            asn1.format = SandwichAPI::encoding_format::ASN1EncodingFormat::ENCODING_FORMAT_PEM;
            cert
        });

        tls_server.set_private_key({
            let mut pkey = SandwichAPI::PrivateKey::new();
            let mut asn1 = pkey.mut_field_static();
            asn1.mut_data().set_filename(PEM_KEY_PATH.to_string());
            asn1.format = SandwichAPI::encoding_format::ASN1EncodingFormat::ENCODING_FORMAT_PEM;
            pkey
        });

        sandwich::Context::try_from(&conf)
    }

    /// Creates the I/O.
    fn create_ios() -> (super::IOBufferRC, super::IOBufferRC) {
        let client = std::rc::Rc::new(std::cell::RefCell::new(super::IOBuffer(
            Vec::<u8>::new(),
            None,
        )));
        let server = std::rc::Rc::new(std::cell::RefCell::new(super::IOBuffer(
            Vec::<u8>::new(),
            None,
        )));

        client.as_ref().borrow_mut().1 = Some(server.clone());
        server.as_ref().borrow_mut().1 = Some(client.clone());
        (client, server)
    }

    /// Test a simple tunnel
    #[test]
    fn test() {
        type Buf = Vec<u8>;
        let mut client_ctx = create_client().unwrap();
        let mut server_ctx = create_server().unwrap();

        let (client_io, server_io) = create_ios();

        let mut client_tun =
            sandwich::Tunnel::new(&mut client_ctx, &mut *client_io.as_ref().borrow_mut()).unwrap();
        let mut server_tun =
            sandwich::Tunnel::new(&mut server_ctx, &mut *server_io.as_ref().borrow_mut()).unwrap();

        {
            let e = client_tun.handshake();
            assert!(e.is_err());
            assert_eq!(
                e.unwrap_err().what(),
                SandwichProto::HandshakeState::HANDSHAKESTATE_WANT_READ
            );
        }

        {
            let e = server_tun.handshake();
            assert!(e.is_err());
            assert_eq!(
                e.unwrap_err().what(),
                SandwichProto::HandshakeState::HANDSHAKESTATE_WANT_READ
            );
        }

        {
            let e = client_tun.handshake();
            assert!(e.is_ok());
        }

        {
            let e = server_tun.handshake();
            assert!(e.is_ok());
        }

        {
            let op = client_tun.write(&PING_MSG);
            assert!(op.is_ok());
            assert_eq!(op.unwrap(), PING_MSG.len());
        }

        {
            let mut buf = Buf::new();
            buf.resize(PING_MSG.len(), 0u8);
            let op = server_tun.read(&mut buf);
            assert!(op.is_ok());
            assert_eq!(op.unwrap(), PING_MSG.len());
            assert_eq!(&buf, &PING_MSG);
        }

        {
            let op = server_tun.write(&PONG_MSG);
            assert!(op.is_ok());
            assert_eq!(op.unwrap(), PONG_MSG.len());
        }

        {
            let mut buf = Buf::new();
            buf.resize(PONG_MSG.len(), 0u8);
            let op = client_tun.read(&mut buf);
            assert!(op.is_ok());
            assert_eq!(op.unwrap(), PONG_MSG.len());
            assert_eq!(&buf, &PONG_MSG);
        }
    }
}
